xapian-core/tests/unittest.cc

```diff
@@ -881,6 +881,49 @@ static bool test_parseunsigned1()
     return true;
 }
 
+template<typename S>
+inline static void parsesigned_helper() {
+    S val;
+    const S max_val = numeric_limits<S>::max();
+    const S min_val = numeric_limits<S>::min();
+    tout << "Testing with parsesigned_helper" << endl;
+    TEST(parse_signed("0", val));
+    TEST_EQUAL(val, 0);
+    TEST(parse_signed("99", val));
+    TEST_EQUAL(val, 99);
+    TEST(parse_signed("-99", val));
+    TEST_EQUAL(val, -99);
+    TEST(parse_signed(str(max_val).c_str(), val));
+    TEST_EQUAL(val, max_val);
+    TEST(parse_signed(str(min_val).c_str(), val));
+    TEST_EQUAL(val, min_val);
+    TEST(!parse_signed("", val));
+    TEST(!parse_signed("abc", val));
+    TEST(!parse_signed("0a", val));
+    TEST(!parse_signed("-99a", val));
+    TEST(!parse_signed("-a99", val));
+    TEST(!parse_signed("--99", val));
+
+    unsigned long long temp = max_val + 1ull;
+    TEST(!parse_signed(str(temp).c_str(), val));
+
+    if (sizeof(S) < sizeof(long long)) {
+	long long temp1 = static_cast<long long>(min_val) - 1LL;
```

Comment:
There are warnings in many of the builds due to this expression overflowing, which the compiler is able to spot that at compile time.  We avoid executing it which avoids triggering the undefined behaviour, but the compiler warnings need to be avoided even when they don't actually indicate a genuine problem (because if the build is warning clean we can enable `-Werror` in developer builds to stop warnings creeping in).

It occurs to me we can convert the negated value and then prepend `-` as a string op, which would avoid such warnings, and also would allow this test to be done for any type.  Something like:

```
unsigned long long temp1 = -min_val + 1ull;
TEST(!parse_signed(("-" + str(temp1)).c_str(), val));
```

The names `temp` and `temp1` aren't usefully descriptive - something like `one_too_large` and `one_too_small` (or perhaps `one_too_small_negated` for my tweaked version) would make it clearer to the reader what this code is testing.

Reply:
fixed in a fixup commit